### DISCUSSION TITLE
Distinguishing between update/insert in upsert

### DISCUSSION THREAD

**Discussion by fchu** [prisma/prisma#3432] [Q&A] [answered]
Hi there,
I couldn't find a clear answer in the docs to the following question:
Since the call return the updated/modified object, how does one know whether an upsert was an update or a create? 
Thank you!

**Comment by nikolasburk** [comment]
Hey @fchu, good question!

One approach to resolve could be to compare the ID that you're passing to the `where` option in `upsert` with the ID that's returned by the `upsert` query:

```ts
const maybeUserId = 42
const user = await prisma.user.upsert({
  create: {
    name: "Nikolas",
    email: "burk@prisma.io"
  },
  update: {
    name: "Nikolas",
    email: "burk@prisma.io",
  },
  where: {
    id: maybeUserId
  }
})
if (user.id === maybeUserId) {
  console.log(`An existing user was updated`)
} else {
  console.log(`A new user was created`)
}
```

Let me know if you have any further questions, happy to help 🙂

**Comment by fchu** [comment]
Ah yes, great answer for a typical use case!

However I failed to mention that the id itself is provided by the upsert call, not generated (think email, or auth id), in which case I suppose there is no solution?

**Reply by ryands17** [reply, marked as answer]
There's another way to do this as well. Consider this example:

```prisma
model User {
  id        Int      @default(autoincrement()) @id
  name      String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

If `createdAt` is equal to `updatedAt`, it's a create, else t's an upsert.

**Reply by fchu** [reply]
That's a great idea! Thank you!

**Reply by HartS** [reply]
@ryands17 is this `@updatedAt` attribute still supported? If so, it's not documented anywhere, not really, for Prisma 2.

**Reply by HartS** [reply]
There's a note here on updating from prisma 1, but `@updatedAt` attribute isn't documented anywhere else as far as google shows me: https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-postgres#workarounds-2

**Reply by janpio** [reply]
See my reply on your issue: https://github.com/prisma/prisma/issues/6512#issuecomment-817353499

**Reply by englut** [reply]
I'm noticing that in the latest version of Prisma 3.4.2, the updatedAt & createdAt no longer line up, it's off by 0.001 second.

**Reply by mziwisky** [reply]
this method won't work if you're using `upsert` just to ensure the record exists because `updatedAt` won't change, so the two timestamps might still match even if the record already existed.  e.g.

```prisma
model User {
  id        Int      @default(autoincrement()) @id
  name      String
  email     String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

```js
prisma.user.create({
  data: { name: 'Jon', email: 'jon@jon.jon' }
})

prisma.user.upsert({
  where: { email: 'jon@jon.jon' },
  create: { name: 'Jon', email: 'jon@jon.jon' },
  update: {},
})
```

the result from `upsert` will be a record with matching timestamps, even though it wasn't just created.

**Reply by okomarov** [reply]
@mziwisky Is this because of the `update: {}`? Semantically I read that operation as "force create".

**Reply by mziwisky** [reply]
@okomarov i'm not certain what you mean by "force create".  i think of it more like "create if doesn't already exist". but my point is there's no generic way to tell whether an upsert just created the record, or just returned the existing one.

**Reply by okomarov** [reply]
I misread your comment.

However, if `updateAt` and `createdAt` are the same, wouldn't that be the expected result, since it falls back onto get-or-create behaviour when the update is empty (`update: {}`)?

**Reply by luixal** [reply]
This doesn't seem to work for me. Most cases it works out, but sometimes createdAt and updatedAt are slighty different when creating new records:

```js
{
    id: 50,
    userName: 'Resident_99999675',
    account: '00016608',
    panelSerialNumber: '00040AAA',
    createdAt: 2022-07-07T10:51:04.483Z, // here both dates are equal
    updatedAt: 2022-07-07T10:51:04.483Z
  },
  {
    id: 51,
    userName: 'Resident_20016955',
    account: '00003424',
    panelSerialNumber: '00040BBB',
    createdAt: 2022-07-07T10:51:04.489Z, // here they're not
    updatedAt: 2022-07-07T10:51:04.490Z
  },
```

If I add the `@default(now())` to the `updatedAt` in prisma schema like this:

```prisma
updatedAt         DateTime @default(now()) @updatedAt
```

it does work perfectly.

@mziwisky in my case, I'm using upsert and it does work.

**Reply by okomarov** [reply]
@luixal I thought the same at first too and have the same model fields.

Then added some tests and every now and then these tests failed. The occurrence is like 1/100 on CI runs, so very very low, and almost never in local.

**Reply by luixal** [reply]
Then... the only way to truly know if record was created or updated is to handle this logic manually by find and update or create?

**Reply by ghost** [reply]
I've fixed the `createdAt` and `updatedAt` which sometimes can be different in the range of 5 ms

isNewRecord = `updatedAt.getTime() - createdAt.getTime() < 10`

so if the difference in time is less than 10ms it will return true

**Comment by mziwisky** [comment]
Does it fall back onto get-or-create when update is empty?  I guess I wouldn't expect that.  I'd think it would still "touch" the record, e.g. for a model with an `@updateAt` column, which prisma automatically sets to the current time whenever the record is updated, I'd expect it to get updated in this case.  But it does not.  Therefore the record was neither "updated" nor "inserted", so "upsert" doesn't quite seem fitting for what just happened.

**Reply by okomarov** [reply]
Good thoughts and I think ultimately is about expectations as different database engines might differ in this aspect, eg the INSERT ... ON CONFLICT with the default to DO NOTHING.

**Comment by yscumc** [comment]
It would be great if the return object could provide the action taken. I think it would solve many headaches for something that (at least based on my ignorance) doesn't seem like a difficult thing to pass back, besides picking a unique-enough object key that wouldn't collide with real world data.

In my case, I'm using upsert as an insert-if-not-exist operation (suggested [in issue #5815](https://github.com/prisma/prisma/discussions/5815#discussioncomment-404334)) and there's also no update timestamp column. I'm also using a secondary unique key which is not auto-generated in the upsert's `where` clause, so I cannot use that value to compare against the return object. So as far as I can tell, there's no officially supported way for me to be able to differentiate an update vs a create.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
